src/MessageDispatch.KurrentDB/KurrentDbSubscriber.cs

```diff
@@ -0,0 +1,426 @@
+    /// <param name="dispatcher">Dispatcher.</param>
+    /// <param name="logger">Logger.</param>
+    /// <param name="checkpointFilePath">Path of the checkpoint file.</param>
+    /// <returns>A new KurrentDbSubscriber object.</returns>
+    // ReSharper disable once UnusedMember.Global
+    public static KurrentDbSubscriber CreateCatchupSubscriptionSubscribedToAllUsingCheckpoint(
+        KurrentDBClient kurrentDbClient,
+        IDispatcher<ResolvedEvent> dispatcher,
+        ILogger logger,
+        string checkpointFilePath)
+        => new KurrentDbSubscriber(
+                kurrentDbClient,
+                dispatcher,
+                logger,
+                AllStreamName,
+                checkpointFilePath);
+
+    /// <summary>
+    /// Start the subscriber.
+    /// </summary>
+    // ReSharper disable once MemberCanBePrivate.Global
+    public async void Start()
+    {
+        _cts = new CancellationTokenSource();
+
+        while (true)
+        {
+            try
+            {
+                var subscription = CreateSubscription();
+                _logger.LogInformation("Subscribed to '{StreamName}'", _streamName);
+
+                await foreach (var message in subscription.Messages)
+                {
+                    switch (message)
+                    {
+                        case StreamMessage.Event(var @event):
+                            ProcessEvent(@event);
+
+                            var lastProcessedEventPosition = GetLastProcessedPosition(@event);
+
+                            if (_liveOnly && _lastProcessedEventPosition is null)
+                            {
+                                _startingPosition = lastProcessedEventPosition;
+                            }
+
+                            _lastProcessedEventPosition = lastProcessedEventPosition;
+                            break;
+                        case StreamMessage.AllStreamCheckpointReached(var allPosition):
+                            _lastProcessedEventPosition = allPosition.CommitPosition;
+                            WriteCheckpoint((ulong)_lastProcessedEventPosition);
+                            break;
+                        case StreamMessage.CaughtUp:
+                            _logger.LogInformation("Stream caught up: {0}", _streamName);
+                            IsLive = true;
+                            break;
+                        case StreamMessage.FellBehind:
+                            _logger.LogWarning("Stream falling behind: {0}", _streamName);
+                            IsLive = false;
+                            break;
+                    }
+                }
+            }
+            // User initiated drop, do not resubscribe
+            catch (OperationCanceledException ex)
+            {
+                IsLive = false;
+                _logger.LogInformation(ex, "Event Store subscription dropped {0}", SubscriptionDroppedReason.Disposed);
```

Comment:
Should this be `SubscriptionDroppedReason.Cancelled`? Otherwise these two catch statements could be combined

Reply:
Looks like this is probably resolvable now, I think Rich mentioned this was primarily taken from their docs.